strawberryfields/apps/train/parametrization.py

```diff
@@ -0,0 +1,363 @@
+    >>> cov = A_to_cov(a)
+
+    Args:
+        A (array): the adjacency matrix
+
+    Returns:
+        array: the covariance matrix of :math:`A`
+    """
+    n = len(A)
+    A_big = np.block([[A, 0 * A], [0 * A, np.conj(A)]])
+    I = np.identity(2 * n)
+    X = Xmat(n)
+    return np.linalg.inv(I - X @ A_big) - I / 2
+
+
+# TODO: update embedding class in type hint
+class VGBS:
+    r"""Create a variational GBS model for optimization and machine learning.
+
+    An input adjacency matrix :math:`A` can be varied using:
+
+    .. math::
+
+        A(\theta) = W(\theta) A W(\theta)
+
+    with :math:`W` a diagonal matrix of weights that depend on a set of parameters :math:`\theta`.
+
+    By varying :math:`\theta`, the distribution of samples from GBS can be preferentially biased
+    towards solving stochastic optimization and unsupervised machine learning problems.
+
+    The above variational model can be used in both the threshold and PNR modes of GBS, which
+    is specified using the ``threshold`` flag. An initial value for the mean number of clicks
+    (if ``threshold=True``) or mean number of photons (if ``threshold=False``) is also required.
+    The initial value ``n_mean`` should be set high since varying :math:`\theta` can only lower
+    the mean number of clicks or photons.
+
+    The mapping from :math:`\theta` to :math:`W(\theta)` is specified using the ``embedding``
+    argument.
+
+    **Example usage:**
+
+    >>> g = nx.erdos_renyi_graph(4, 0.7, seed=1967)
+    >>> a = nx.to_numpy_array(g)
+    >>> embedding = Exp()  # TODO does this need updating?
+    >>> vgbs = VGBS(a, 3, embedding, True)
+    >>> params = np.array([0.05, 0.1, 0.02, 0.01])
+    >>> vgbs.A(params)
+    array([[0.        , 0.86070798, 0.93239382, 0.94176453],
+           [0.86070798, 0.        , 0.88692044, 0.89583414],
+           [0.93239382, 0.88692044, 0.        , 0.97044553],
+           [0.94176453, 0.89583414, 0.97044553, 0.        ]])
+    >>> vgbs.n_mean(params)
+    1.8906000271819803
+    >>> vgbs.generate_samples(params, 2)
+    array([[0, 0, 1, 1],
+           [1, 1, 0, 0]])
+
+    Args:
+        A (array): the input adjacency matrix :math:`A`
+        n_mean (float): the initial mean number of clicks or photons
+        embedding: the method of converting from trainable parameters :math:`\theta` to
+            :math:`W(\theta)` #TODO: any update from other PR?
+        threshold (bool): determines whether to use GBS in threshold or PNR mode
+        samples (array): an optional array of samples from :math:`A` used to speed up gradient
+            calculations #TODO: more info
+    """
+    def __init__(
+        self,
+        A: np.ndarray,
+        n_mean: float,
+        embedding,
+        threshold: bool,
+        samples: Optional[np.ndarray] = None,
+    ):
+        if not np.allclose(A, A.T):
+            raise ValueError("Input must be a NumPy array corresponding to a symmetric matrix")
+        self.A_init = A
```

Comment:
is `A_init` vs `A` really helpful? I understand it's meant to clarify that it's the input `A`, but I get confused with many `A` flying around

Reply:
This actually touches open a point we need to get right: the role of `A` and `A_scaled`. I feel like there should only be one `A`, the rescaled one. So for example, if we initialize the class with an unnormalized `A` and then call `vgbs.A(np.ones(dim))` (so `W` is the identity), we should just get back the rescaled `A`.

So essentially the rescaled `A` is always the point of reference; we just make it easy for the user to input even an unnormalized one. Alternatively, we could force the user to always input a rescaled matrix

Reply:
Great! I have updated to:

- Remove the `_WAW` method
- Make the `A` method be the one that gives the scaled WAW
- Remove the `A_scaled` method as this is now `A`

One thing to note is that there is that there is still an `A_init` attribute - it would be great to have an `A` attribute _and_ and `A` method, but this is not possible in Python(?).